openfisca_core/populations.py

```diff
@@ -0,0 +1,565 @@
+# -*- coding: utf-8 -*-
+
+import traceback
+
+from typing import Iterable
+
+import numpy as np
+
+from openfisca_core.entities import Role
+from openfisca_core.indexed_enums import EnumArray
+from openfisca_core.holders import Holder
+
+
+ADD = 'add'
+DIVIDE = 'divide'
+
+
+def projectable(function):
+    """
+    Decorator to indicate that when called on a projector, the outcome of the function must be projected.
+    For instance person.household.sum(...) must be projected on person, while it would not make sense for person.household.get_holder.
+    """
+    function.projectable = True
+    return function
+
+
+class Population(object):
+    def __init__(self, entity):
+        self.simulation = None
+        self.entity = entity
+        self._holders = {}
+        self.count = 0
+        self.ids = []
+
+    def clone(self, simulation):
+        result = Population(self.entity)
+        result.simulation = simulation
+        result._holders = {variable: holder.clone(result) for (variable, holder) in self._holders.items()}
+        result.count = self.count
+        result.ids = self.ids
+        return result
+
+    def empty_array(self):
+        return np.zeros(self.count)
+
+    def filled_array(self, value, dtype = None):
+        return np.full(self.count, value, dtype)
+
+    def __getattr__(self, attribute):
+        projector = get_projector_from_shortcut(self, attribute)
+        if not projector:
+            raise AttributeError("You tried to use the '{}' of '{}' but that is not a known attribute.".format(attribute, self.entity.key))
+        return projector
+
+    def get_index(self, id):
+        return self.ids.index(id)
+
+    # Calculations
+
+    def check_array_compatible_with_entity(self, array):
+        if not self.count == array.size:
+            raise ValueError("Input {} is not a valid value for the entity {} (size = {} != {} = count)".format(
+                array, self.key, array.size, self.count))
+
+    def check_period_validity(self, variable_name, period):
+        if period is None:
+            stack = traceback.extract_stack()
+            filename, line_number, function_name, line_of_code = stack[-3]
+            raise ValueError('''
+You requested computation of variable "{}", but you did not specify on which period in "{}:{}":
+    {}
+When you request the computation of a variable within a formula, you must always specify the period as the second parameter. The convention is to call this parameter "period". For example:
+    computed_salary = person('salary', period).
+See more information at <https://openfisca.org/doc/coding-the-legislation/35_periods.html#periods-in-variable-definition>.
+'''.format(variable_name, filename, line_number, line_of_code).encode('utf-8'))
```

Comment:
```suggestion
'''.format(variable_name, filename, line_number, line_of_code))
```

(I think the `encode` is legacy from Python 2, and it makes the error less readable in Python 3)